checker/src/org/checkerframework/checker/upperbound/UpperBoundAnnotatedTypeFactory.java

```diff
@@ -0,0 +1,770 @@
+        super(checker);
+        UNKNOWN = AnnotationUtils.fromClass(elements, UpperBoundUnknown.class);
+
+        valueAnnotatedTypeFactory = getTypeFactoryOfSubchecker(ValueChecker.class);
+        minLenAnnotatedTypeFactory = getTypeFactoryOfSubchecker(MinLenChecker.class);
+        addAliasedAnnotation(IndexFor.class, createLTLengthOfAnnotation(new String[0]));
+        addAliasedAnnotation(IndexOrHigh.class, createLTEqLengthOfAnnotation(new String[0]));
+        this.postInit();
+    }
+
+    @Override
+    protected ExpressionAnnotationHelper createExpressionAnnotationHelper() {
+        List<Class<? extends Annotation>> annos = new ArrayList<>();
+        annos.add(LTLengthOf.class);
+        annos.add(LTEqLengthOf.class);
+        annos.add(IndexFor.class);
+        annos.add(IndexOrHigh.class);
+        annos.add(LTOMLengthOf.class);
+        return new ExpressionAnnotationHelper(this, annos) {
+            @Override
+            public TreeAnnotator createExpressionAnnotationTreeAnnotator(
+                    AnnotatedTypeFactory factory) {
+                return new ExpressionAnnotationTreeAnnotator(factory, this) {
+                    @Override
+                    public Void visitMemberSelect(MemberSelectTree tree, AnnotatedTypeMirror type) {
+                        // UpperBoundTreeAnnotator changes the type of array.length to @LTEL
+                        // ("array"). If the ExpressionAnnotationTreeAnnotator tries to viewpoint
+                        // adapt it based on the declaration of length; it will fail.
+                        if (tree.getIdentifier().contentEquals("length")
+                                && InternalUtils.typeOf(tree.getExpression()).getKind()
+                                        == TypeKind.ARRAY) {
+                            return null;
+                        }
+                        return super.visitMemberSelect(tree, type);
+                    }
+                };
+            }
+        };
+    }
+
+    @Override
+    public AnnotationMirror aliasedAnnotation(AnnotationMirror a) {
+        if (AnnotationUtils.areSameByClass(a, IndexFor.class)) {
+            List<String> stringList =
+                    AnnotationUtils.getElementValueArray(a, "value", String.class, true);
+            return createLTLengthOfAnnotation(stringList.toArray(new String[0]));
+        }
+        if (AnnotationUtils.areSameByClass(a, IndexOrHigh.class)) {
+            List<String> stringList =
+                    AnnotationUtils.getElementValueArray(a, "value", String.class, true);
+            return createLTEqLengthOfAnnotation(stringList.toArray(new String[0]));
+        }
+        return super.aliasedAnnotation(a);
+    }
+
+    /**
+     * Queries the MinLen Checker to determine if there is a known minimum length for the array. If
+     * not, returns null.
+     */
+    public Integer minLenFromExpressionTree(ExpressionTree tree) {
```

Comment:
Since 0 is a valid minlen, making this return 0 when there is no minlen erases that distinction - which its callers make use of. However, -1 isn't a valid minlen, so I'm going to do as you suggest, but with -1 instead of 0.